src/steadystateproblem.cpp

```diff
@@ -206,21 +207,16 @@ void SteadystateProblem::applyNewtonsMethod(ReturnData *rdata, Model *model,
             /* Check residuals vs tolerances */
             converged = wrms < RCONST(1.0);
 
-            if (converged) {
-                /* Ensure positivity of the found state */
-                for (ix = 0; ix < model->nx_solver; ix++) {
-                    if (x[ix] < 0.0) {
-                        x[ix] = 0.0;
-                        converged = FALSE;
-                    }
-                }
-            } else {
-                /* increase dampening factor (superfluous, if converged) */
-                gamma = fmin(1.0, 2.0 * gamma);
+            if (!converged && newtonSolver->dampingFactorMode==NewtonDampingFactorMode::on) {
+              /* increase dampening factor (superfluous, if converged) */
+              gamma = fmin(1.0, 2.0 * gamma);
             }
-        } else {
-            /* Reduce dampening factor */
+        } else if (newtonSolver->dampingFactorMode==NewtonDampingFactorMode::on) {
+            /* Reduce dampening factor and raise an error when becomes too small */
             gamma = gamma / 4.0;
+            if (gamma < 1e-6)
```

Comment:
How sure are we about this number?
Might it be good to make this an user-adjustable parameter?

Reply:
Up to you. It is just a reasonable lower bound. Usually, you do not need to go lower. If you need, then most likely you will not converge at all.

Reply:
I see the point, but I like avoiding magic numbers which are hard-coded in algorithms...
As the damping factor is something like a step size, I'm wondering whether linking it to the error tolerances might be a good idea...

Reply:
Agreed with the magic numbers. Maybe for start just add them as function arguments with the given default values. Linking to error tolerances seems like a start of a new project. Sounds good, but maybe too much here.

Reply:
Totally up to you.

Reply:
Let's make it a user option for the moment.

Reply:
OK

Reply:
Is 1e-6 a reasonable default value or you prefer a different number?

Reply:
Seems reasonable to me. Either that or 1e-8, which is our default for relative tolerances... Both are fine for me.